### PR TITLE
Add Tversky and Focal Tversky Losses

### PR DESCRIPTION
**Pull request by charleygros**
#### Original papers:
- [Tversky Loss 2017](https://arxiv.org/pdf/1706.05721.pdf)
- [Tversky Focal Loss 2018](https://arxiv.org/pdf/1810.07842.pdf)

#### Motivation:
Have been recommended for high class imbalance issues.

#### Limitation:
Some hyperparameters to optimise.

#### Background:
- Tversky Index is a asymmetric similarity measure
- If alpha=beta=0.5 --> Tversky index is equivalent to the Dice Score
- If alpha=beta=1 --> Tversky index is equivalent to the Jaccard Index
- alpha > beta --> penalise false negatives
- Focal Tversky Loss --> Same idea than Focal Loss: non-linear nature of the loss gives control over how the loss behaves at different values of the Tversky index obtained.
- For gamma > 1: the loss is higher for samples where Tversky < 0.5 --> the model focus on harder samples.

### CODE REVIEW

File: ivadomed/losses.py [hunk 1]
```diff
@@ -114,3 +114,103 @@ def forward(self, input, target):
...
+        """
+        # Compute TP
+        tp = torch.sum(y_true * y_pred)
+        # Compute FN
+        fn = torch.sum(y_true * (1 - y_pred))
+        # Compute FP
+        fp = torch.sum((1 - y_true) * y_pred)
+        # Compute Tversky for the current class, see Equation 3 of the original paper
+        numerator = tp + self.smooth
+        denominator = tp + self.alpha * fn + self.beta * fp + self.smooth
```
Comment (olix86) [inline ivadomed/losses.py:161]: It's a detail, but I think the alpha and beta are switched up, from the paper : 
> where α and β control the magnitude of penalties for FPs and FNs, respectively.

Comment (olix86) [reply ivadomed/losses.py:161]: The confusion probably comes from `eq3` since they define :
> the `p0i` is the probability of voxel `i` be a lesion and `p1i` is the probability of voxel `i` be a non-lesion

Which is kind of the opposite of what I would have expected

Comment (charleygros) [reply ivadomed/losses.py]: Mmm this is confusing I agree. I ve just swapped!

File: ivadomed/losses.py [hunk 2]
```diff
@@ -114,3 +114,103 @@ def forward(self, input, target):
...
+        # TODO: Add class_of_interest?
+        for i_label in range(n_classes):
+            # Get samples for a given class
+            y_pred, y_true = input[:, i_label, ], target[:, i_label, ]
+            # Compute Tversky index
+            tversky_index = self.tversky.tversky_index(y_pred, y_true)
+            # Compute Focal Tversky loss, Equation 4 in the original paper
+            focal_tversky_sum += torch.pow(1 - tversky_index, exponent=1 / self.gamma)
+
+        # TODO: Take the opposite?
```
Comment (andreanne-lemay) [inline ivadomed/losses.py:215]: Here what do you mean with this TODO :)?

Comment (charleygros) [reply ivadomed/losses.py:215]: We have the convention to minimize our loss during training, ie -1=good, 0=bad.
I just want to make sure that I have not made a mess of the signs --> I will carefully check this via #257 and make sure all losses have the same trend.

File: ivadomed/losses.py [hunk 3]
```diff
@@ -114,3 +114,103 @@ def forward(self, input, target):
...
+        denominator = tp + self.alpha * fp + self.beta * fn + self.smooth
+        tversky_label = numerator / denominator
+        return tversky_label
+
+    def forward(self, input, target):
+        n_classes = input.shape[1]
+        tversky_sum = 0.
+
+        # TODO: Add class_of_interest?
+        for i_label in range(n_classes):
```
Comment (andreanne-lemay) [inline ivadomed/losses.py:170]: I am wondering here: why do we have to loop? The operations in tversky_index (substraction, addition and division) can be done on multichannel images, no?

Comment (charleygros) [reply ivadomed/losses.py:170]: Because in the FocalTverskyLoss --> we apply the gamma to (1-tversky_class_i) --> we need to compute them per label --> that's why I created this loop.
I could re-write this with array manipulation if you think it would be better.
